java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -0,0 +1,131 @@
+ *       family: Text,
+ *       limit: Number
+ *     } = new PetStats {
+ *       ratio: new Domestication {
+ *         trained: input.cats,
+ *         total: input.cats + input.dogs,
+ *       },
+ *       family: input.household,
+ *       limit: 5
+ *     }
+ *   ```
+ *
+ *   This can be translated to:
+ *
+ *   ```
+ *   DependencyNode.AggregateValue(
+ *    "ratio" to DependencyNode.AggregateValue(
+ *      "trained" to DependencyNode.PrimitiveValue("input", "cats"),
+ *      "total" to DependencyNode.DerivedFrom(
+ *        DependencyNode.PrimitiveValue("input", "cats"),
+ *        DependencyNode.PrimitiveValue("input", "dogs")
+ *      )
+ *    ),
+ *    "family" to DependencyNode.PrimitiveValue("input", "household"),
+ *    "limit" to DependencyNode.LITERAL
+ *   )
+ *   ```
+ *
+ *   We can represent this DAG graphically:
+ *
+ *                    (root)
+ *            __________|______
+ *           /      |          \
+ *          /      /        _(ratio)_
+ *        /       /        |         \
+ *   (limit)  (family)  (trained)  (total)
+ *     |        |          |     /   |
+ *     |        |          |   /     |
+ *     x     (household)  (cats)  (dogs)
+ *
+ *   This internal representation, in turn, can be translated into the following claims:
+ *
+ *   ```
+ *   claim output.ratio.trained derives from input.cats
+ *   claim output.ratio.total derives from input.cats and derives from input.dogs
+ *   claim output.family derives from input.household
+ *   ```
+ */
+sealed class DependencyNode {
+
+  /** An unmodified input (from a handle connection) used in a Paxel [Expression]. */
+  data class PrimitiveValue(val path: Path = emptyList()) : DependencyNode() {
+    constructor(vararg fields: Identifier) : this(listOf(*fields))
+  }
+
+  /** Represents modification of an access path within a Paxel [Expression]. */
+  data class DerivedFrom(val inputs: Set<DependencyNode> = emptySet()) : DependencyNode() {
+
+    constructor() : this(emptySet())
+
+    constructor(vararg paths: Path) : this(paths.map { PrimitiveValue(it) }.toSet())
+
+    constructor(vararg nodes: DependencyNode) : this(flatten(*nodes))
+
+    companion object {
+      /** Flatten nested sets of [DependencyNode]s.*/
+      private fun flatten(vararg nodes: DependencyNode): Set<DependencyNode> {
+        return nodes.flatMap { node ->
+          when (node) {
+            is PrimitiveValue -> setOf(node)
+            is DerivedFrom -> node.inputs
+            else -> throw IllegalArgumentException(
+              "Nodes must be a 'PrimitiveValue' or 'DerivedFrom'."
+            )
+          }
+        }.toSet()
+      }
+    }
+  }
+
+  /** Associates [Identifier]s with [DependencyNode]s. */
+  data class AggregateValue(
+    val associations: Map<Identifier, DependencyNode> = emptyMap()
+  ) : DependencyNode() {
+
+    constructor(vararg pairs: Pair<Identifier, DependencyNode>) : this(pairs.toMap())
```

Comment:
A brief comment?

Reply:
Should I comment all the ctors?

Reply:
For some constructors, it will be obvious. If not, adding a brief comment would help. I will let you make the judgement.